config/desc/imsim/instcat.py

```diff
@@ -0,0 +1,627 @@
+        logger.debug("HA = %s",self.meta['HA'])
+
+    def getAirmass(self, altitude):
+        """
+        Function to compute the airmass from altitude using equation 3
+        of Krisciunas and Schaefer 1991.
+
+        Parameters
+        ----------
+        altitude: float
+            Altitude of pointing direction in degrees.
+
+        Returns
+        -------
+        float: the airmass in units of sea-level airmass at the zenith.
+        """
+        altRad = np.radians(altitude)
+        return 1.0/np.sqrt(1.0 - 0.96*(np.sin(0.5*np.pi - altRad))**2)
+
+    def get(self, field):
+        if field not in self.meta:
+            raise ValueError("OpsimMeta field %s not present in instance catalog"%field)
+        return self.meta[field]
+
+    def getHourAngle(self, mjd, ra):
+        """
+        Compute the local hour angle of an object for the specified
+        MJD and RA.
+
+        Parameters
+        ----------
+        mjd: float
+            Modified Julian Date of the observation.
+        ra: float
+            Right Ascension (in degrees) of the object.
+
+        Returns
+        -------
+        float: hour angle in degrees
+        """
+        # cf. http://www.ctio.noao.edu/noao/content/coordinates-observatories-cerro-tololo-and-cerro-pachon
+        lsst_lat = '-30d 14m 40.68s'
+        lsst_long = '-70d 44m 57.90s'
+        lsst_elev = '2647m'
+        lsst_loc = astropy.coordinates.EarthLocation.from_geodetic(
+                        lsst_lat, lsst_long, lsst_elev)
+
+        time = astropy.time.Time(mjd, format='mjd', location=lsst_loc)
+        # Get the local apparent sidereal time.
+        last = time.sidereal_time('apparent').degree
+        ha = last - ra
+        return ha
+
+def OpsimMeta(config, base, value_type):
+    """Return one of the meta values stored in the instance catalog.
+    """
+    meta = galsim.config.GetInputObj('opsim_meta_dict', config, base, 'OpsimMeta')
+
+    req = { 'field' : str }
+    opt = { 'num' : int }  # num, if present, was used by GetInputObj
+    kwargs, safe = galsim.config.GetAllParams(config, base, req=req, opt=opt)
+    field = kwargs['field']
+
+    val = value_type(meta.get(field))
+    return val, safe
+
+def InstCatObj(config, base, ignore, gsparams, logger):
+    """Build an object according to info in instance catalog.
+    """
+    inst = galsim.config.GetInputObj('instance_catalog', config, base, 'InstCat')
+
+    # Setup the indexing sequence if it hasn't been specified.
+    # The normal thing with a catalog is to just use each object in order,
+    # so we don't require the user to specify that by hand.  We can do it for them.
+    galsim.config.SetDefaultIndex(config, inst.getNObjects())
+
+    req = { 'index' : int }
+    opt = { 'num' : int }
+    kwargs, safe = galsim.config.GetAllParams(config, base, req=req, opt=opt)
+    index = kwargs['index']
+
+    rng = galsim.config.GetRNG(config, base, logger, 'InstCatObj')
+    bp = base['bandpass']
```

Comment:
This line seems to be looking for a `bandpass` entry in the header part of the instance catalogs, but the current instance catalogs use `filter` (as an integer) to specify the band.   If I add a `bandpass u` entry to my instcat, I get a `ValueError` from the attempt to convert to `float` in line 427.

Mike: How does this work when you run?   I assume you've modified your instcat files, but how?

Reply:
Never mind about this one.  I was using an old version of the `imsim.yaml` file.  When I update to the current version, I don't encounter this problem.

Reply:
To clarify, it's looking for the 'bandpass' key in the base config dict, which is a fully constructed `galsim.Bandpass` object.

The path is:
1. This line converts "filter" in the instance catalog into a "band", which is just a single letter that things in the config file can use.
```
self.meta['band'] = 'ugrizy'[self.meta['filter']]
```
2. The config specifies the bandpass of the image as
```
bandpass: { type: OpsimMetaBandpass }
```
3. The OpsimMetaBandpass class builds the galsim.Bandpass as:
```
band = meta.get('band')
bandpass = galsim.Bandpass('LSST_%s.dat'%band, wave_type='nm')
bandpass = bandpass.withZeropoint('AB')
```
4. GalSim saves the current bandpass in use at the top level so anyone can easily access it.  So somewhere in the galsim code it saves the return value from OpsimMetaBandpass as `base['bandpass']`, which this line is then able to use.